File: packages/nvim-pack/nvim_pack.lua
local request_path = assert(vim.env.NVIM_PACK_REQUEST, "NVIM_PACK_REQUEST is not set")
local response_path = assert(vim.env.NVIM_PACK_RESPONSE, "NVIM_PACK_RESPONSE is not set")

---@param path string
---@return any
local function read_json(path)
	local lines = vim.fn.readfile(path, "b")
	return vim.json.decode(table.concat(lines, "\n"))
end

---@param path string
---@param value any
local function write_json(path, value)
	local status = vim.fn.writefile({ vim.json.encode(value) }, path, "b")
	if status ~= 0 then
		error("could not write nvim-pack response")
	end
end

---@param values string[]
---@return string[]
local function sorted(values)
	table.sort(values)
	return values
end

---@return string[]
local function active_plugin_names()
	local names = {}
	for _, plugin in ipairs(vim.pack.get(nil, { info = false })) do
		if plugin.active then
			names[#names + 1] = plugin.spec.name
		end
	end
	return sorted(names)
end

---@param requested string[]
---@return string[] selected
---@return string[] selection_errors
local function select_active_plugins(requested)
	local active = active_plugin_names()
	if #requested == 0 then
		return active, {}
	end

	local active_set = {}
	for _, name in ipairs(active) do
		active_set[name] = true
	end

	local selected, selection_errors = {}, {}
	for _, name in ipairs(requested) do
		if active_set[name] then
			selected[#selected + 1] = name
		else
			selection_errors[#selection_errors + 1] = name
		end
	end
	return selected, selection_errors
end

---@param lines string[]
---@return string[] details
---@return string[] updates
---@return string[] errors
---@return table<string, string> targets
local function parse_confirmation(lines)
	local details, updates, errors, targets = {}, {}, {}, {}
	local group, current_update = nil, nil
	local in_updates = false

	for _, line in ipairs(lines) do
		local next_group = line:match("^# (%S+)")
		if next_group then
			if in_updates and next_group ~= "Update" then
				in_updates = false
			end
			group = next_group
			in_updates = group == "Update"
		else
			if in_updates then
				details[#details + 1] = line
			end
		end

		local name = line:match("^## (.+)$")
		if name then
			name = name:gsub(" %(not active%)$", "")
			current_update = group == "Update" and name or nil
			if group == "Update" then
				updates[#updates + 1] = name
			elseif group == "Error" then
				errors[#errors + 1] = name
			end
		end

		local target = line:match("^Revision after:%s+(%x+)")
		if target and current_update then
			targets[current_update] = target
		end
	end

	while details[1] == "" do
		table.remove(details, 1)
	end
	while details[#details] == "" do
		table.remove(details)
	end
	return details, updates, errors, targets
end

---@param selected string[]
---@param offline boolean
---@return table
local function inspect_updates(selected, offline)
	if offline then
		vim.pack.update(selected, { offline = true })
	else
		vim.pack.update(selected)
	end

	local bufnr = vim.api.nvim_get_current_buf()
	if vim.bo[bufnr].filetype ~= "nvim-pack" then
		error("vim.pack.update did not create its confirmation buffer")
	end
	local lines = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
	local details, updates, errors, targets = parse_confirmation(lines)
	return {
		bufnr = bufnr,
		details = details,
		errors = errors,
		targets = targets,
		updates = updates,
	}
end

---@param selected string[]
---@param lock_path string
---@return table
local function snapshot_check_state(selected, lock_path)
	local lock_existed = vim.fn.filereadable(lock_path) == 1
	local lock_before = lock_existed and vim.fn.readfile(lock_path, "b") or {}
	local origins, errors = {}, {}

	for _, plugin in ipairs(vim.pack.get(selected, { info = false })) do
		local result = vim.system({ "git", "remote", "get-url", "origin" }, {
			cwd = plugin.path,
			text = true,
		}):wait()
		if result.code == 0 then
			origins[plugin.spec.name] = {
				path = plugin.path,
				url = vim.trim(result.stdout),
			}
		else
			errors[#errors + 1] = plugin.spec.name
		end
	end

	return {
		errors = errors,
		lock_before = lock_before,
		lock_existed = lock_existed,
		origins = origins,
	}
end

---@param snapshot table
---@param lock_path string
local function restore_check_state(snapshot, lock_path)
	for name, origin in pairs(snapshot.origins) do
		local result = vim.system({ "git", "remote", "set-url", "origin", origin.url }, {
			cwd = origin.path,
			text = true,
		}):wait()
		if result.code ~= 0 then
			error("could not restore origin for " .. name .. ": " .. vim.trim(result.stderr))
		end
	end

	local lock_exists_after = vim.fn.filereadable(lock_path) == 1
	local lock_after = lock_exists_after and vim.fn.readfile(lock_path, "b") or {}
	if snapshot.lock_existed == lock_exists_after and vim.deep_equal(snapshot.lock_before, lock_after) then
		return
	end

	local status
	if snapshot.lock_existed then
		status = vim.fn.writefile(snapshot.lock_before, lock_path, "b")
	else
		status = vim.fn.delete(lock_path)
	end
	if status ~= 0 then
		error("could not restore nvim-pack-lock.json after check")
	end
end

---@param selected string[]
---@param report table
---@param lock_path string
---@return string[]
local function verify_applied_updates(selected, report, lock_path)
	local installed = {}
	for _, plugin in ipairs(vim.pack.get(selected, { info = false })) do
		installed[plugin.spec.name] = plugin.rev
	end

	local lock_ok, lock_lines = pcall(vim.fn.readfile, lock_path)
	local json_ok, lock = pcall(vim.json.decode, lock_ok and table.concat(lock_lines, "\n") or "")
	local apply_errors = {}
	for _, name in ipairs(report.updates) do
		local entry = json_ok and lock.plugins and lock.plugins[name] or nil
		local target = report.targets[name]
		if not target or installed[name] ~= target or not entry or entry.rev ~= target then
			apply_errors[#apply_errors + 1] = name
		end
	end
	return apply_errors
end

---@param selected string[]
---@param lock_path string
---@return table
local function check_updates(selected, lock_path)
	local snapshot = snapshot_check_state(selected, lock_path)
	if #snapshot.errors > 0 then
		return { errors = snapshot.errors }
	end

	local check_ok, report = xpcall(function()
		return inspect_updates(selected, false)
	end, debug.traceback)
	local restore_ok, restore_error = xpcall(function()
		restore_check_state(snapshot, lock_path)
	end, debug.traceback)

	if not check_ok then
		local suffix = restore_ok and "" or ("\nCleanup also failed: " .. restore_error)
		error(report .. suffix)
	end
	if not restore_ok then
		error(restore_error)
	end
	return report
end

---@param selected string[]
---@param lock_path string
---@param offline boolean
---@return table
local function apply_updates(selected, lock_path, offline)
	local report = inspect_updates(selected, offline)
	if #report.errors == 0 and #report.updates > 0 then
		vim.api.nvim_buf_call(report.bufnr, function()
			vim.cmd.write()
		end)
		report.apply_errors = verify_applied_updates(selected, report, lock_path)
	end
	return report
end

---@param request table
---@return table
local function update(request)
	local selected, selection_errors = select_active_plugins(request.plugins or {})
	local empty_report = {
		apply_errors = {},
		details = {},
		errors = {},
		selection_errors = selection_errors,
		updates = {},
	}
	if #selection_errors > 0 or #selected == 0 then
		return empty_report
	end

	local lock_path = vim.fs.joinpath(vim.fn.stdpath("config"), "nvim-pack-lock.json")
	local report
	if request.apply then
		report = apply_updates(selected, lock_path, request.offline == true)
	else
		report = check_updates(selected, lock_path)
	end
	return {
		apply_errors = report.apply_errors or {},
		details = report.details or {},
		errors = report.errors or {},
		selection_errors = selection_errors,
		updates = report.updates or {},
	}
end

---@param dry_run boolean
---@return table
local function prune(dry_run)
	local orphans = {}
	for _, plugin in ipairs(vim.pack.get(nil, { info = false })) do
		if plugin.active == false then
			orphans[#orphans + 1] = plugin.spec.name
		end
	end
	sorted(orphans)
	if not dry_run and #orphans > 0 then
		vim.pack.del(orphans)
	end
	return { names = orphans }
end

---@param request table
---@return table
local function dispatch(request)
	if request.command == "list" then
		return { names = active_plugin_names() }
	end
	if request.command == "prune" then
		return prune(request.dry_run == true)
	end
	if request.command == "update" then
		return update(request)
	end
	error("unknown nvim-pack command: " .. tostring(request.command))
end

local request = read_json(request_path)
local ok, response = xpcall(dispatch, debug.traceback, request)
if not ok then
	response = { fatal_error = response }
end
write_json(response_path, response)

File: packages/nvim-pack/nvim_pack.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import json
import os
import shutil
import subprocess
import sys
import tempfile
from collections.abc import Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import cast


class NvimPackError(Exception):
    pass


@dataclass(frozen=True)
class ListArgs:
    pass


@dataclass(frozen=True)
class CheckArgs:
    plugins: tuple[str, ...]


@dataclass(frozen=True)
class UpdateArgs:
    plugins: tuple[str, ...]
    select: bool


@dataclass(frozen=True)
class PruneArgs:
    dry_run: bool


Args = ListArgs | CheckArgs | UpdateArgs | PruneArgs


@dataclass(frozen=True)
class UpdateReport:
    updates: list[str]
    errors: list[str]
    apply_errors: list[str]
    selection_errors: list[str]
    details: list[str]


def add_plugin_arguments(parser: argparse.ArgumentParser) -> None:
    """Add optional active-plugin selection shared by check and update."""
    _ = parser.add_argument(
        "plugins",
        nargs="*",
        metavar="PLUGIN",
        help="active plugin names to process (default: all active plugins)",
    )


def create_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="nvim-pack",
        description="Inspect and maintain plugins managed by Neovim's vim.pack.",
    )
    subparsers = parser.add_subparsers(dest="command", required=True)

    list_parser = subparsers.add_parser(
        "list",
        help="list active plugin names",
        description="List active plugin names, one per line.",
    )
    list_parser.set_defaults(command="list")

    check_parser = subparsers.add_parser(
        "check",
        help="show available updates",
        description="Fetch and show available updates without applying them.",
    )
    add_plugin_arguments(check_parser)

    update_parser = subparsers.add_parser(
        "update",
        help="apply available updates",
        description="Apply available updates and rewrite nvim-pack-lock.json.",
    )
    add_plugin_arguments(update_parser)
    _ = update_parser.add_argument(
        "--select",
        action="store_true",
        help="choose from updatable plugins with fzf",
    )

    prune_parser = subparsers.add_parser(
        "prune",
        help="remove orphaned plugins",
        description=(
            "Remove plugins that are installed but no longer added by the current "
            "Neovim config."
        ),
    )
    _ = prune_parser.add_argument(
        "--dry-run",
        action="store_true",
        help="list orphaned plugins without deleting them",
    )
    return parser


def parse_args(argv: list[str] | None = None) -> Args:
    namespace = create_parser().parse_args(argv)
    command = cast(str, namespace.command)
    if command == "list":
        return ListArgs()
    if command in {"check", "update"}:
        plugins = tuple(dict.fromkeys(cast(list[str], namespace.plugins)))
        if command == "check":
            return CheckArgs(plugins)
        select = cast(bool, namespace.select)
        if select and plugins:
            create_parser().error("update --select cannot be combined with plugin names")
        return UpdateArgs(
            plugins=plugins,
            select=select,
        )
    if command == "prune":
        return PruneArgs(dry_run=cast(bool, namespace.dry_run))
    raise AssertionError(f"unhandled command: {command}")


def require_nvim() -> None:
    """Fail when the host Neovim needed to load the user's config is absent."""
    if shutil.which("nvim") is None:
        raise NvimPackError(
            "nvim was not found on PATH; nvim-pack uses the host's own nvim so "
            + "the active plugin set matches this machine's config"
        )


def lua_script_path() -> Path:
    """Return the packaged Lua adapter, with a sibling fallback for source runs."""
    configured = os.environ.get("NVIM_PACK_SCRIPT")
    path = Path(configured) if configured else Path(__file__).with_suffix(".lua")
    if not path.is_file():
        raise NvimPackError(f"nvim-pack Lua adapter was not found: {path}")
    return path


def string_list(response: dict[str, object], key: str) -> list[str]:
    """Read a required string-list field from a Lua response."""
    if key not in response:
        raise NvimPackError(f"nvim-pack Lua response omitted required field {key!r}")
    value = response[key]
    if not isinstance(value, list):
        raise NvimPackError(f"nvim-pack Lua response field {key!r} was not a string list")

    strings: list[str] = []
    for item in cast(list[object], value):
        if not isinstance(item, str):
            raise NvimPackError(f"nvim-pack Lua response field {key!r} was not a string list")
        strings.append(item)
    return strings


def run_nvim(request: dict[str, object]) -> dict[str, object]:
    """Execute the Lua adapter after normal config startup and return its response."""
    with tempfile.TemporaryDirectory(prefix="nvim-pack-") as temporary_directory:
        directory = Path(temporary_directory)
        request_path = directory / "request.json"
        response_path = directory / "response.json"
        _ = request_path.write_text(json.dumps(request), encoding="utf-8")

        environment = os.environ.copy()
        environment.update(
            {
                "NVIM_PACK_REQUEST": str(request_path),
                "NVIM_PACK_RESPONSE": str(response_path),
                "NVIM_PACK_SCRIPT": str(lua_script_path()),
            }
        )
        try:
            completed = subprocess.run(
                [
                    "nvim",
                    "--headless",
                    "-c",
                    "lua dofile(vim.env.NVIM_PACK_SCRIPT)",
                    "-c",
                    "qa!",
                ],
                check=False,
                capture_output=True,
                text=True,
                errors="replace",
                env=environment,
            )
        except OSError as error:
            raise NvimPackError(f"could not run nvim: {error}") from error

        if not response_path.is_file():
            detail = completed.stderr.strip() or f"exit status {completed.returncode}"
            raise NvimPackError(f"nvim did not complete the vim.pack command: {detail}")
        try:
            response = cast(object, json.loads(response_path.read_text(encoding="utf-8")))
        except (OSError, json.JSONDecodeError) as error:
            raise NvimPackError(f"could not read nvim-pack Lua response: {error}") from error

    if not isinstance(response, dict):
        raise NvimPackError("nvim-pack Lua response was not an object")
    response_object = cast(dict[object, object], response)
    if not all(isinstance(key, str) for key in response_object):
        raise NvimPackError("nvim-pack Lua response had a non-string key")
    typed_response = {cast(str, key): value for key, value in response_object.items()}
    fatal_error = typed_response.get("fatal_error")
    if fatal_error is not None:
        if not isinstance(fatal_error, str):
            raise NvimPackError("nvim-pack Lua fatal_error was not a string")
        raise NvimPackError(f"nvim-pack Lua failed: {fatal_error}")
    return typed_response


def query_updates(
    apply: bool,
    plugins: Sequence[str],
    *,
    offline: bool = False,
) -> UpdateReport:
    """Return a validated update report for selected active plugins."""
    if offline and not apply:
        raise ValueError("offline mode is only valid when applying updates")
    response = run_nvim(
        {
            "apply": apply,
            "command": "update",
            "offline": offline,
            "plugins": list(plugins),
        }
    )
    report = UpdateReport(
        updates=string_list(response, "updates"),
        errors=string_list(response, "errors"),
        apply_errors=string_list(response, "apply_errors"),
        selection_errors=string_list(response, "selection_errors"),
        details=string_list(response, "details"),
    )
    if report.selection_errors:
        names = ", ".join(report.selection_errors)
        raise NvimPackError(f"not active vim.pack plugins: {names}")
    if report.errors:
        names = ", ".join(report.errors)
        raise NvimPackError(f"could not check updates for: {names}")
    if report.apply_errors:
        names = ", ".join(report.apply_errors)
        raise NvimPackError(
            f"updates did not finish cleanly for: {names}; plugin state may be partial"
        )
    return report


def select_updates(names: Sequence[str]) -> list[str]:
    """Interactively choose update candidates with fzf; cancellation is empty."""
    fzf = shutil.which("fzf")
    if fzf is None:
        raise NvimPackError("fzf was not found on PATH; update --select requires fzf")
    try:
        completed = subprocess.run(
            [
                fzf,
                "--multi",
                "--prompt=Update plugins> ",
                "--header=Tab: toggle selection | Enter: update | Esc: cancel",
            ],
            input="".join(f"{name}\n" for name in names),
            stdout=subprocess.PIPE,
            text=True,
            errors="replace",
            check=False,
        )
    except OSError as error:
        raise NvimPackError(f"could not run fzf: {error}") from error
    if completed.returncode in {1, 130}:
        return []
    if completed.returncode != 0:
        raise NvimPackError(f"fzf failed with exit status {completed.returncode}")

    allowed = set(names)
    selected = list(dict.fromkeys(completed.stdout.splitlines()))
    if any(name not in allowed for name in selected):
        raise NvimPackError("fzf returned a plugin that was not an update candidate")
    return selected


def format_update_result(names: list[str], apply: bool) -> str:
    if not names:
        return "All selected vim.pack plugins are up to date."
    noun = "plugin" if len(names) == 1 else "plugins"
    verb = "Can update" if not apply else "Updated"
    return f"{verb} {len(names)} {noun}: {', '.join(names)}"


def format_prune_result(names: list[str], dry_run: bool) -> str:
    if not names:
        return "vim.pack is in sync; no orphaned plugins to prune."
    noun = "plugin" if len(names) == 1 else "plugins"
    verb = "Would prune" if dry_run else "Pruned"
    return f"{verb} {len(names)} orphaned {noun}: {', '.join(names)}"


def run_update(
    apply: bool,
    plugins: Sequence[str],
    *,
    offline: bool = False,
) -> int:
    """Check or apply updates for selected active plugins."""
    report = query_updates(apply, plugins, offline=offline)
    print(format_update_result(report.updates, apply))
    if not apply and report.details:
        print()
        print("\n".join(report.details))
    return 0


def run_selected_update() -> int:
    """Check all plugins, select candidates with fzf, and apply from fetched refs."""
    candidates = query_updates(False, ()).updates
    if not candidates:
        print(format_update_result([], False))
        return 0

    selected = select_updates(candidates)
    if not selected:
        print("No plugins selected; nothing updated.")
        return 0
    return run_update(True, selected, offline=True)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv)
    try:
        require_nvim()
        if isinstance(args, ListArgs):
            names = string_list(run_nvim({"command": "list"}), "names")
            if names:
                print("\n".join(names))
            return 0
        if isinstance(args, CheckArgs):
            return run_update(False, args.plugins)
        if isinstance(args, UpdateArgs):
            return run_selected_update() if args.select else run_update(True, args.plugins)
        response = run_nvim(
            {
                "command": "prune",
                "dry_run": args.dry_run,
            }
        )
        print(format_prune_result(string_list(response, "names"), args.dry_run))
        return 0
    except NvimPackError as error:
        print(f"error: {error}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
